### PR TITLE
[WNMGDS-1826] Upgrade to React 17

### PR DESCRIPTION
**Pull request by zarahzachz**
## Summary

[Jira ticket](https://jira.cms.gov/browse/WNMGDS-1826)

- Upgrade project from React 16 to 17
- Remove React imports from functional components when possible
- Fix typos I find

Removing React import to [improve performance](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#whats-a-jsx-transform):
> “With the new transform, you can use JSX without importing React.
> 
> Depending on your setup, its compiled output may slightly improve the bundle size.”

However, when I run `yarn build` it looks like our bundle size _increased_ 20KiB (to 492 KiB). I don't know if our measurements are off or if there's something wrong here, but based on what I'm reading this should've helped reduce bundle size.

There are also several components I couldn't remove the import from. 

The following is a list of `class` components I found that rely on the React import:
- AccordionItem
- Autocomplete
- Choice
- ChoiceList
- DateInput
- Dropdown
- Select
- FilterChip
- FormLabel
- Table
- Mask
- TextField
- NavigationMenu (mgov)

The following are functional components, but rely on several React methods like `Children`, `cloneElement`, etc. to function and removing the React import didn't seem to provide benefits:
- TableHead
- TableRow
- Tabs
- LabelMask

### CODE REVIEW

File: babel.config.js [hunk 1]
```diff
@@ -2,7 +2,12 @@ module.exports = function (api) {
   api.cache(true);
 
   const presets = [
-    '@babel/preset-react',
+    [
```
Comment (zarahzachz) [inline babel.config.js:5]: From [React 17 docs](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#manual-babel-setup)

File: .eslintrc.js [hunk 2]
```diff
@@ -1,5 +1,10 @@
 module.exports = {
-  extends: ['eslint:recommended', 'plugin:react/recommended', 'plugin:react-hooks/recommended'],
+  extends: [
+    'eslint:recommended',
+    'plugin:react/recommended',
+    'plugin:react-hooks/recommended',
+    'plugin:react/jsx-runtime',
```
Comment (zarahzachz) [inline .eslintrc.js:6]: From [RTL docs](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/react-in-jsx-scope.md)

**Comment by zarahzachz** [comment]
React 17 changes `onBlur` and `onFocus` appear to take place under the hood and [exists to support IE 9-11](https://blog.saeloun.com/2021/05/14/react-17-uses-browse-focusin-focusout-for-onfocus-onblur). Doesn't appear to impact our use of these events.

**Comment by zarahzachz** [comment]
I've tested this update in React versions [16](https://github.cms.gov/CMS-WDS/taxtools), [17](https://github.cms.gov/CMS-WDS/wds-accounts-healthcare), and [18](https://github.cms.gov/CMS-WDS/learn-frontend-gatsby) and all experiments have passed 🎉

File: packages/design-system/src/components/Accordion/Accordion.stories.jsx [hunk 3]
```diff
@@ -25,7 +23,7 @@ Default.args = {
     <AccordionItem key="1" heading="First Amendment" contentClassName="first-amendment" defaultOpen>
       <p>
         We the People of the United States, in Order to form a more perfect Union, establish
-        Justice, insure domestic Tranquility, provide for the common defence, promote the general
+        Justice, insure domestic Tranquility, provide for the common defense, promote the general
```
Comment (pwolfert) [inline packages/design-system/src/components/Accordion/Accordion.stories.jsx:26]: This is in the original text of the Constitution 😛. There are a few other little spelling-variation morsels in the Article I text that we have in another story.

File: packages/design-system/src/components/Drawer/DrawerToggle.tsx [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import Button, { ButtonProps } from '../Button/Button';
-import React, { useEffect, useRef } from 'react';
+import { useEffect, useRef } from 'react';
```
Comment (pwolfert) [inline packages/design-system/src/components/Drawer/DrawerToggle.tsx:2]: I like that the magic part of React is going away but that we're still explicit about importing the functions that we explicitly use.

File: packages/docs/src/styles/components/responsiveExample.scss [hunk 5]
```diff
@@ -1,4 +1,4 @@
-.c-resposive-example__example-wrapper {
+.c-responsive-example__example-wrapper {
```
Comment (pwolfert) [inline packages/docs/src/styles/components/responsiveExample.scss:1]: Nice 👍

File: packages/docs/src/helpers/loremIpsumGenerator.ts [hunk 6]
```diff
@@ -3,7 +3,7 @@ type LoremIpsumTextLength = 's' | 'm' | 'l';
 const lorem = {
   s: 'We the People of the United States',
   m: 'We the People of the United States, in Order to form a more perfect Union',
-  l: 'We the People of the United States, in Order to form a more perfect Union, establish Justice, insure domestic Tranquility, provide for the common defence, promote the general Welfare, and secure the Blessings of Liberty to ourselves and our Posterity, do ordain and establish this Constitution for the United States of America.',
+  l: 'We the People of the United States, in Order to form a more perfect Union, establish Justice, insure domestic Tranquility, provide for the common defense, promote the general Welfare, and secure the Blessings of Liberty to ourselves and our Posterity, do ordain and establish this Constitution for the United States of America.',
```
Comment (pwolfert) [inline packages/docs/src/helpers/loremIpsumGenerator.ts:6]: Here it is again. Don't think we need to correct it, since it's original.

**Review by pwolfert — COMMENTED** [review]
Here are my comments so far. I'm getting some errors running Gatsby locally, which I'm still looking into

**Comment by pwolfert** [comment]
Regarding the missing `React` in Gatsby, I've tried nuking all `node_modules` and `.cache` and rebuilding, and I also don't see any other babel config files. Not sure why it's not working the same way as the other packages.

**Comment by zarahzachz** [comment]
Looked into the Gatsby issues and it appears the Gatsby team made the [JSX transforms support disabled by default](https://github.com/gatsbyjs/gatsby/issues/28657), which explains the errors. 

Looking at the Gatsby docs, there's [a piece of code I can add to our Babel config](https://www.gatsbyjs.com/docs/reference/release-notes/v3.0/#babel-preset-gatsby) to fix this.

**Comment by zarahzachz** [comment]
> Looking at the Gatsby docs, there's [a piece of code I can add to our Babel config](https://www.gatsbyjs.com/docs/reference/release-notes/v3.0/#babel-preset-gatsby) to fix this.

JK, this doesn't work either.

Circling back to the solution proposed in the GH issue linked above. That works and the site builds fine.

File: packages/design-system/src/components/DateField/MultiInputDateField.stories.jsx [hunk 7]
```diff
@@ -1,4 +1,3 @@
-import React, { useState } from 'react';
```
Comment (forrestbaer) [inline packages/design-system/src/components/DateField/MultiInputDateField.stories.jsx:1]: need to put `useState` back, seeing some errors in storybook

File: packages/design-system/src/components/DateField/SingleInputDateField.stories.jsx [hunk 8]
```diff
@@ -1,4 +1,3 @@
-import React, { useState } from 'react';
```
Comment (forrestbaer) [inline packages/design-system/src/components/DateField/SingleInputDateField.stories.jsx:1]: Same here, `useState` needed.

**Review by forrestbaer — COMMENTED** [review]
I can't get gatsby to build, maybe the fix hasn't been commited yet? Clearing node modules and .cache I can get storybook to build but unable to get Gatsby going. Nice work, love the code removals.

**Review by forrestbaer — APPROVED** [review]
This is looking good now, build working.  There will be some things to resolve in my build changes so I'm not putting up a PR for that until this is merged and I have a chance to go through build testing there.

**Pull request merged**